hal/mbed_sleep_manager.c

```diff
@@ -19,13 +19,116 @@
 #include "sleep_api.h"
 #include "mbed_error.h"
 #include <limits.h>
+#include <stdio.h>
 
 #if DEVICE_SLEEP
 
 // deep sleep locking counter. A target is allowed to deep sleep if counter == 0
 static uint16_t deep_sleep_lock = 0U;
 
-void sleep_manager_lock_deep_sleep(void)
+#ifdef MBED_SLEEP_TRACING_ENABLED
+
+// Length of the identifier extracted from the driver name to store for logging.
+#define IDENTIFIER_WIDTH 7
+// Number of drivers that can be stored in the structure
+#define STATISTIC_COUNT  10
+
+typedef struct sleep_statistic {
+    char identifier[IDENTIFIER_WIDTH];
+    uint8_t count;
+} sleep_statistic_t;
+
+static sleep_statistic_t sleep_stats[STATISTIC_COUNT];
+
+static const char* strip_path(const char* const filename)
```

Comment:
All this runtime processing can be avoided, if we use proper compile time pre-processing macros (`__MODULE__` for ARM and `__BASE_FILE__` for GCC_ARM). Not blocking the release, but please update this in next PR.

Reply:
That's fine for ARM and GCC, but the only way to do it for IAR seems to be enabling `--no_path_in_file_macros` which seems excessive unless it can be applied selectively to this file.

Reply:
--no_path_in_file_macros : I guess we will be adding this anyways in future for unified tracing.

Reply:
Instead of using filename can we use function name of the caller( the caller will pass `__FUNCTION__` ), that will avoid all these processing as well, does that work?

Reply:
I think using `__BASE_FILE__` would work better, it can be added when this is altered to use the tracing. 

Using `__FUNCTION__` might be problematic. The names of the functions that call lock and unlock are going to be different so tracking which unlock is associated with each lock would be complicated, multiple drivers are also likely to have the same function names that lock/unlock the sleep manager, and it's difficult to identify which driver locked the sleep manager based on the function name alone